api/traces/v1/trace_rbac.go

```diff
@@ -0,0 +1,182 @@
+package v1
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/grafana/tempo/pkg/tempopb"
+	commonv1 "github.com/grafana/tempo/pkg/tempopb/common/v1"
+	tracev1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+
+	apilogsv1 "github.com/observatorium/api/api/logs/v1"
+)
+
+func WithTraceQLNamespaceSelect() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			query := r.URL.Query()
+			q := query.Get("q")
+			traceQL, err := url.QueryUnescape(q)
+			if err != nil {
+				next.ServeHTTP(w, r)
+				return
+			}
+			if traceQL == "" {
+				traceQL = "{ }"
+			}
+			traceQL = traceQL + " | select(resource.k8s.namespace.name)"
+			query.Set("q", traceQL)
+			r.URL.RawQuery = query.Encode()
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func responseRBACModifier(log log.Logger) func(response *http.Response) error {
+	return func(response *http.Response) error {
+		allowedNamespaces := map[string]bool{}
+		namespaces := apilogsv1.AllowedNamespaces(response.Request.Context())
+		for _, ns := range namespaces {
+			allowedNamespaces[ns] = true
+		}
+		level.Debug(log).Log("AllowedNamespaces", allowedNamespaces)
+
+		if strings.Contains(response.Request.URL.Path, "/api/traces/") || strings.Contains(response.Request.URL.Path, "/api/search") {
+			if response.StatusCode == http.StatusOK {
+				// Uncompressed reader
+				var reader io.ReadCloser
+				var err error
+
+				// Read what Jaeger UI sent back (which might be compressed)
+				switch response.Header.Get("Content-Encoding") {
+				case "gzip":
+					reader, err = gzip.NewReader(response.Body)
+					if err != nil {
+						return err
+					}
+					defer reader.Close()
+				case "deflate":
+					reader = flate.NewReader(response.Body)
+					defer reader.Close()
+				default:
+					reader = response.Body
+				}
+
+				b, err := io.ReadAll(reader)
+				if err != nil {
+					return err
+				}
+				strResponse := string(b)
+				buf := bytes.NewBufferString(strResponse)
+				response.Header["Content-Length"] = []string{fmt.Sprint(buf.Len())}
```

Comment:
The new, modified response (after `traceRBAC()`) can have a different length than the original response